libwacom/libwacom.c

```diff
@@ -283,15 +285,24 @@ get_device_info (const char            *path,
 	}
 
 	*name = g_strdup (g_udev_device_get_sysfs_attr (device, "name"));
-	/* Try getting the name from the parent if that fails */
-	if (*name == NULL) {
+	*fw = g_strdup (g_udev_device_get_sysfs_attr (device, "fw_name"));
+	/* Try getting the name/fw_name from the parent if that fails */
+	if (*name == NULL || *fw == NULL) {
 		GUdevDevice *parent;
-
 		parent = g_udev_device_get_parent (device);
-		if (!parent)
-			goto out;
-		*name = g_strdup (g_udev_device_get_sysfs_attr (parent, "name"));
+
+		while (parent && *name == NULL && *fw == NULL) {
```

Comment:
Is it possible that `name != NULL` but `fw == NULL` at this point? Should this be `while (parent && (*name == NULL || *fw == NULL)) {`?

Reply:
yep, good point. fixed. still untested but less buggy now :)